Implementation of "Accumulation Bounds" in SymPy

gxyd opened the pull request:
For information regarding "What Limits is?" see commit message.
Please bear with me, i am not very good at writing `OOP` in `Python right now. But will try to improve upon this.
ping @jksuom

Commit:
Implementation of "AccumBounds" in SymPy

Note
====

We started with a working name of `Limits` but after thinking and suggestions
from @jksuom it was succesfully changed to `AccumulationBounds`(alias
`AccumBounds`), since this name better represents the actual functionality.

Currently limit(sin(x), x, oo) returns nan which is not good since it
looses information that the actual value is bounded by -1 and 1. Since
that information is quite useful and to store that i following along
lines as suggested by @jksuom. On [this
comment](https://github.com/sympy/sympy/issues/9934#issuecomment-145452969)
on issue #9934. Wolfram also does give this information by saying -1 to
1 for sin(oo).

Problems currently
==================
sin(oo) is kept un-evaluated hence
>>> sin(oo) - sin(oo)
0

which is wrong. Using AccumBounds
>>> sin(oo)
AccumBounds(-1, 1)    # here the bounds of sin i.e <-1, 1> is returned
>>> sin(oo) - sin(oo)
AccumBounds(-2, 2)    # this is actual expected answer, since the value
                oscillates between -2 and 2.

Note: AccumBounds is not necessarily going to return exact bound for a
function, even Wolfram does not always return the exact bound. Since
it depends upon how the user expresses the input(i.e whether in
simplified form).

For ex.
>>> sin(oo)*cos(oo)
AccumBounds(-1, 1)          # this may be returned

I am not saying this will be returned but as i implement it further
let's see how good we(with help of @jksuom) be able to go with its
Implementation.
`sin(oo)*cos(oo)` can be written as `sin(2*x)/2)`
>>> sin(2*x)/2
AccumBounds(-1/2, 1/2)     # then exact bounds are returned

Hargup's code of Interval Arithmeic included
Part of code is derived from @hargup 's PR

Division 1/AccumBounds corrected

trignometry AccumBounds used

start -> min and end -> max for AccumBounds

test changed made for incorporating sin(oo) to AccumBounds(-1, 1)

tests for oo and AccumBounds added

gxyd commented:
I made some changes in the division of `Limits` here

gxyd commented on sympy/functions/elementary/trigonometric.py:
```diff
@@ -450,21 +469,17 @@ def fdiff(self, argindex=1):
-                # https://github.com/sympy/sympy/issues/5196
-                # For now, we return un-evaluated.
-                return
+                return Limits(-1, 1)
```
I will add a note here, mentioning why it is better to return `Limits(-1, 1)` instead of returning `nan`.

gxyd commented:
Currently `sin(1) + Limits(-2, 3)` does not evaluate (simply returned as it is). While `Limits(-2, 3) + sin(1)` evaluated correctly to `Limits(-2 + sin(1), 3 + sin(1))`. Probably `sin` does not move `radd` to it?

jksuom commented:
Perhaps you should set ``_op_priority`` to 11 or bigger. See the comment in expr.py line 94.

gxyd commented on sympy/series/tests/test_limits.py:
```diff
@@ -205,6 +209,20 @@ def test_doit():
+    #assert limit(tan(k)/3 - tan(k + 1), k, oo) == Limits(-oo, oo)
+
+    # test for issue #9934
+    assert limit(simplify(Sum(cos(n).rewrite(exp), (n, 0, k)).doit().rewrite(sin)), k, oo) == \
```
Perhaps this fails right now, i will get to this. Currently this returns:
```
cos(1) + [-3, 1]
────────────────
2⋅(-1 + cos(1)) 
```

gxyd replied:
@jksuom Can you please look why this is not working? The simplifcation of
```
>>> cos(1) + Limits(-3, 1)
<cos(1) - 3, cos(1) + 1>    # this works correctly
```

But this does not work
```
>>> limit(simplify(Sum(cos(n).rewrite(exp), (n, 0, k)).doit().rewrite(sin)), k, oo)
cos(1) + [-3, 1]
────────────────           # here the simplification of cos(1) + <-3, 1> does not occur.
2⋅(-1 + cos(1)) 
```

Commit:
printing of Limits using $$\lrangle \rangle$$

tests for printing added

Pretty printing of Limits corrected and tests added

duplicate test removed and start->min and end->max

gxyd commented on sympy/calculus/util.py:
```diff
@@ -112,3 +117,316 @@ def elm_domain(expr, intrvl):
+    Note
+    ====
+
+    The main focus in the interval arithmetic is on the simplest way to calculate
```
@asmeurer I had added this note for what `Limits` is supposed to do and what it does not do.

gxyd commented:
I have introduced different(from `Intervals`) printing of `Limits`, now `Limits` uses `<1, 2>` for printing. Since those angular brackets would indicate that we are not dealing with an interval (in fact, often only with a part of it).

gxyd commented:
ping @jksuom @asmeurer I have updated the docstring for the `Limits` and made changed in division of `Limits`. Still one thing remains that
```
>>> limit(simplify(Sum(cos(n).rewrite(exp), (n, 0, k)).doit().rewrite(sin)), k, oo)
(cos(1) + <-3, 1>)/(2*(-1 + cos(1)))
```
This does not get simplified.

jksuom commented:
``cos(1) + <-3, 1>`` is not processed? I suspect that the priority mechanism should take care of this. Is it possible to set the ``_op_priority`` of ``Limits``?

gxyd commented:
> cos(1) + <-3, 1> is not processed?

Seems like you missed [this comment](https://github.com/sympy/sympy/pull/10051#discussion-diff-45602383), later on i just commented this test.

> Is it possible to set the _op_priority of Limits?

Ok, let me see for this.

gxyd commented:
To me seems like the general problem is here:
```
>>> Add(cos(1), Limits(1, 2))
cos(1) + Limits(1, 2)           # does not work

>>> cos(1) + Limits(1, 2)
<cos(1) + 1, cos(1) + 2>     # works Ok

>>> Add(1, Limits(1, 2))
Limits(2, 3)                  # works Ok
```
Same behaviour for `Mul`.

jksuom commented:
I am not sure what should be done with Add and Mul. In some cases it might be better to leave a complicated term outside Limits. Also, there could be a separate function for moving the outside terms into Limits if that is desired.

gxyd commented:
can you tell any criterion on the basis of which i should tell whether the outside expression should go inside `Limits`.
I made some modification only on my "local machine".
```
>>> Limits(1, 2)*2
<1, 2>*2
```
Isn't there any criterion which is could use to simplify "this"(above ones') but not complex ones.

gxyd commented:
> In some cases it might be better to leave a complicated term outside Limits.

I don't see anything wrong in taking any `real-expression` inside the `Limits`, as long as any expression is real it can come inside the `Limits`. That's what i think.

jksuom commented:
It is technically quite correct to move any real expressions inside the ``Limits``. My only concern is readability. I think that sometimes ``complicated_real_expression + <a, b>`` might be preferable to ``<complicated_real_expression + a, complicated_read_expression + b>``. So I would not worry too much at this point if some expressions with ``Limits`` are not automatically transformed into that form.

gxyd commented:
> So I would not worry too much at this point if some expressions with Limits are not automatically transformed into that form.

Okay. I have added the tests for corner cases. I guess the PR is ready for review.

jksuom commented on sympy/functions/elementary/tests/test_trigonometric.py:
```diff
@@ -21,9 +22,15 @@ def test_sin():
     assert sin(oo*I) == oo*I
     assert sin(-oo*I) == -oo*I
-    assert sin(oo).args[0] == oo
+    assert sin(oo) == Limits(-1, 1)
```
This seems to be a copy of a line above.

jksuom commented on sympy/functions/elementary/trigonometric.py:
```diff
@@ -222,13 +224,30 @@ def fdiff(self, argindex=1):
+                return Limits(-1, 1)
+
+        if isinstance(arg, Limits):
+            start, end = arg.start, arg.end
```
``start`` and ``end`` are attributes of ``Interval``. As ``Limit`` is now as a subclass of ``Interval`` there might be some confusion. Does this mean that ``min`` and ``max`` should be renamed (again) to be on the safe side?

gxyd commented:
I had forgot to change the `start`, `end` to `min`, `max`. I have changed it in all places. I don't think there is any point to use `start`, `end` if we use `max`, `min`.

jksuom commented:
I have been thinking of the role of ``Interval``. Is there any place where it is essentially used?

gxyd commented:
The `in` operator is there with the `Interval` class, which is essential for `Limits` class. May be the `Limits` class can have its own `in` operator, that will be easy(just comparing with end-points). Also `intersect` is there which is used in the code. Do you mean in these terms usage of `Interval`? then yes.

gxyd commented:
One more thing to notice is that. Since `Limits` is a subclass of `Interval` so `.is_subset(S.Reals)` is `True` for all `Limits`. I think i should set `.is_real` is to `True` for `Limits`, since the actual value lies in `reals` and should be treated as that?

gxyd commented:
I was thinking of adding `comparison` operators for `Limits`. For example
```
>>> Limits(1, 3) < 4
True
>>> Limits(1, 3) < 2
None
>>> Limits(1, 3) > Limits(4, 5)
False
```
And similar behavior. Perhaps that will be useful.

gxyd commented:
For references i am mentioning the issues here, this PR is supposed address #5299, #9934 .
[Edit] I will update this comment with more if i find the ones related to this.

Commit:
`<`, `>`, `<=`, `>=` added for comparison of Limits

Limits no longer a subclass of `Interval`

S.false removed from Basic call for Limits, docstring for relationals

repetitive test removed and Pow approach changed for Limits

comments for future work added

`Limits` -> `AccumulationBounds` (alias `AccumBounds`)
also code from #2682 taken from Harsha's PR.

fixes #5299 #9934

changes `_print_Limits` -> `_print_AccumulationBouds`

jksuom commented:
It seems that code from ``Interval`` is actually in use to some extent. This probably means that the names of end points have to be redefined unless extra code is added. Consider, for example, the ``in`` operator. If that is inherited from ``Limits``, then the end point should be called ``start`` and ``end`` (I think). Or else the values of ``min`` and ``max`` should be copied.

> .is_subset(S.Reals) is True for all Limits. I think i should set .is_real is to True for Limits

I have no definite opinion of whether ``Limits`` should be a subset of reals or not, though I am inclined to say no. . Maybe its significance will only become apparent in practise. On the other hand, perhaps it would be useful to have ``.is_real`` ``True``.

The ``in`` operator is easy to define in ``Limits``, As to ``intersect``, a bit more code would be needed, but I am not sure if the operation would be necessary. It seems to be used only in some tests.

So it probably would not be hard to define ``Limits`` that is independent of ``Interval``. Then even ``.is_real`` might feel more natural.

gxyd commented:
Now `Limits` is no longer a `subclass` of `Interval`. Some changes will be required in `union` only. The use of `union` for `Limits` lie only for the case of `division` of two `Interval` may not necessarily a `Limit`.

gxyd commented on sympy/calculus/tests/test_util.py:
```diff
@@ -31,3 +33,151 @@ def test_not_empty_in():
+    assert Limits(-1, 3)/Limits(3, 4) == Limits(-S(1)/3, 1)
+    assert Limits(-2, 4)/Limits(-3, 4) == Limits(-oo, oo)
+    assert Limits(-3, -2)/Limits(-4, 0) == Limits(S(1)/2, oo)
+#    assert Limits(-3, -2)/Limits(-2, 1) == Union(Limits(-oo, -2), Limits(1, oo))
```
Here is the example where the division of two `Limits` is not a `Limits` object. I will try to see this.

jksuom replied:
``Limits(-3, -2)/Limits(-2, 1)`` yields a TypeError now. I think it should be ``Limits(-oo, oo)``.

gxyd replied:
But a better answer is `Union(Limits(-oo, -2), Limits(1, oo))`. Here is proof of why  -1 ∉ `Limits(-3, -2)/Limits(-2, 1)`.
 Consider this `Limits(-3, -2)` is always `-ve`(or < 0). Hence for `Limits(-3, -2)/Limits(-2, 1)` to attain a value of `-1` the denominator must be `+ve`. So for `Limits(-2, 1)` we should consider only `Limits(0, 1)`(since `Limits(0, 1)` is +ve in this). Now only "minimum" `2` can satisfy `-2/2 = -1`. Hence -1 ∉ `Limits(-3, -2)/Limits(-2, 1)`.

On page 8, Theorem 7 of [this article](http://fab.cba.mit.edu/classes/S62.12/docs/Hickey_interval.pdf). I used that definition of `division` of `Limits`.

jksuom commented:
It seems (at least for the moment) that ``union`` would only be useful in the case where the 'interval' extends over infinity (with -oo and oo essentially identified). Perhaps this could be written in the form ``1/Limits(a, b)`` where ``a < 0 < b``. Another possibility might be ``Limits(1/b, 1/a)`` where the inverted order of max and min would indicate that the complement is intended. (Then the end points could even have the same sign.) But that may make the arithmetic too complicated.

gxyd commented:
> Perhaps this could be written in the form 1/Limits(a, b) where a < 0 < b

For `1/Limits(a, b)` where `a < 0 < b`, the `Limit` is evaluated to `Limits(-oo, oo)` currently(which is correct). So this "would not" work. Also for the second suggestion, you already pointed out the difficulty of doing that. I am also inclined against that.

jksuom commented on sympy/calculus/util.py:
```diff
@@ -112,3 +119,523 @@ def elm_domain(expr, intrvl):
+        if max == min:
+            return max
+
+        return Basic.__new__(cls, min, max, S.false, S.false)
```
It seems that ``S.false`` is not needed any more.

jksuom commented on sympy/calculus/tests/test_util.py:
```diff
@@ -31,3 +33,152 @@ def test_not_empty_in():
+    assert Limits(-2, 0)*(-oo) == Limits(0, oo)
+    assert Limits(-1, 1)*oo == Limits(-oo, oo)
+    assert Limits(-1, 1)*(-oo) == Limits(-oo, oo)
+    assert Limits(-oo, oo) == Limits(-oo, oo)
```
Is something missing here? It does not seem to belong under test_Limits_mul.

gxyd replied:
Yes, i think i something here.

jksuom commented on sympy/calculus/tests/test_util.py:
```diff
@@ -31,3 +33,152 @@ def test_not_empty_in():
+
+
+def test_Limits_subs():
+    assert (x**2 + 2*x + 1).subs(x, Limits(-1, 1)) == Limits(-1, 4)
```
This should probably be inspected more closely. ``x**2 + 2*x + 1`` is non-negative for all real values of ``x``.

jksuom replied:
Maybe this is not too bad. We do not expect exact limits. But this seems worse:
```
>>> ((x + 1)**2).subs(x, Limits(-1, 1))
<4, oo>
```

gxyd replied:
Ya i just saw this. With the latest commit this is fixed.

jksuom commented on sympy/calculus/tests/test_util.py:
```diff
@@ -31,3 +33,152 @@ def test_not_empty_in():
+    assert Limits(1, 2)**(S(5)/2) == Limits(1, 4*sqrt(2))
+    assert Limits(-1, 2)**(S(1)/3) == Limits(-1, 2**(S(1)/3))
+    assert Limits(0, 2)**(S(1)/2) == Limits(0, sqrt(2))
+    assert Limits(0, 2)**(S(1)/2) == Limits(0, sqrt(2))
```
Is this a repetition?

gxyd commented:
Previously
```
>>> Limits(0, 2)**2
<4, oo>
```
Now, that is fixed(this works as follows)
```
>>> Limits(0, 2)**2
<0, 4>
```

jksuom commented on sympy/calculus/util.py:
```diff
@@ -112,3 +119,601 @@ def elm_domain(expr, intrvl):
+
+    `\langle a, b\rangle = \{x \in \mathbb{R} \mid a \le x \le b\}`
+
+    `\langle -\infty, b\rangle = \{x \in \mathbb{R} \mid x \le b\}`
```
I think that the infinities should be included also here. Both of them, in fact. This is because ``Limits`` do not take into account one-sided limits, by design. Hence ``1/0`` should be regarded as denoting both ``-oo`` and ``oo``` (limits on both sides). Consider, for example, the function ``f = 1/(sin(x) - 1 + 1/x)``:
```
>>> limit(1/(sin(x) - 1 + 1/x), x, oo)
<-oo, -1/2>
```
The intended meaning of ``Limits`` is to give an approximate location of the accumulation points of a real function at a limit point. In this case we can also take the points ``x_n = pi/2 + 2*n*pi`` for which we have ``f(x_n) = x_n`` with limit ``oo``. Hence ``oo`` should be included in ``Limits``. This does not mean that any computations need be redefined. This is purely a question of how ``Limits`` are to be interpreted. (We can think of ``-oo`` and ``oo`` as being identified similarly as when they are mapped to ``zoo``.)

jksuom commented on sympy/calculus/util.py:
```diff
@@ -112,3 +119,601 @@ def elm_domain(expr, intrvl):
+    def _intersect(self, other):
+        res = Interval(self.min, self.max).intersect(
+            Interval(other.min, other.max))
+        return Limits(res.min, res.max)
```
I suspect that ``res`` may not always be an ``Interval``. But if it is, then it has ``start`` and ``end``.

gxyd replied:
I think this internal method is no longer required. I will remove this `_intersect` method completely.(we have `intersect` method that works fine.)

jksuom commented on sympy/calculus/util.py:
```diff
@@ -112,3 +119,601 @@ def elm_domain(expr, intrvl):
+                        return Limits(0, oo)
+                    if self.max.is_zero and other.min.is_zero:
+                        return Limits(-oo, 0)
+                    return Limits(-oo, oo)
```
Perhaps there should also be special treatment for the cases where ``other.max.is_zero``.

gxyd replied:
Addressed.

jksuom commented on sympy/calculus/util.py:
```diff
@@ -112,3 +119,601 @@ def elm_domain(expr, intrvl):
+                        if other.max.is_zero:
+                            return Limits(self.max/other.min, oo)
+                        if other.max.is_positive:
+                            return Union(Limits(-oo, self.max/other.max),
```
This should probably be just ``Limits(-oo, oo)`` for now. Maybe something better could be invented later.

gxyd commented:
I think, I have addressed all the comments(but you can make sure). One thing remains
```
>>> Limits(1, 2)**(-oo)
Does not work
```
I will add the code by tomorrow.

jksuom commented on sympy/calculus/util.py:
```diff
@@ -112,3 +119,662 @@ def elm_domain(expr, intrvl):
+
+    `\langle a, b\rangle = \{x \in \mathbb{R} \mid a \le x \le b\}`
+
+    `\langle -\infty, b\rangle = \{x \in \mathbb{R} \cup \{-\infty, \infty\} \mid x \le b\}`
```
Here ``\infty`` should be outside (else it is not included).``-\infty`` could remain, but perhaps the result would look better if both were together at the end.

jksuom replied:
Please note the remark above. The ``\cup \{-\infty, \infty\}`` part should come last, after ``\mid x \le b\}``.
(``x = \infty`` would not be included as it does not satisfy ``x \le b``.

jksuom commented on sympy/calculus/util.py:
```diff
@@ -112,3 +119,672 @@ def elm_domain(expr, intrvl):
+                    if self.max < 1:
+                        return S.Zero
+                    if self.max == 1:
+                        return FiniteSet(0, 1)
```
It looks like this should be ``Limits(0, 1)``.

Consider a function such as ``f = (sin(x) + 1)`` with ``limit(f, x, oo) == Limits(0, 1)``. If ``e`` is another function with ``limit(e, x, oo) == oo``, for example, ``e = x``, then ``limit(f**e, x, oo)`` should be covered by ``Limit(0, 1)**oo``. But for these functions we have ``limit(f**e, x, oo) == Limits(0, 1)``, since ``f**e`` takes repeatedly all values between 0 and 1 in each interval of length ``2*pi`` as ``x`` tends to ``oo``.

This remark also applies below. We should return ``Limits`` instead of ```FiniteSet``s.

jksuom replied:
I am beginning to think that not even ``Limits(0. 1)`` would cover all possibilities (we could have max > 1 in the result).
It might be best to consider only three cases: max < 1 (as above), min > 1 (as below) and else return ``Limits(0, oo)``. The cases with max or min equal to 1 could give almost any results.

gxyd replied:
For the comment above this, i don't agree. `Limits(0, 1)` lies between 0 and 1, so for any value between 0 and 1, raised to the power tending to `oo` is either `0` or `1`. Like
```
>>> limit( (1/2)**x, x, oo)
0
>>> limit( a**x, x, oo)  # for where 0 < a < 1
0
>>> limit( 1**x, x, oo)
1
```

Hence only two values are possible, either 0 or 1, so i think `FiniteSet(0, 1)` should be the answer instead of `Limits(0, 1)`. Did i missed something?

gxyd replied:
Though the example you gave( 1+ sin(x)) is not exactly `Limits(0, 1)`, but is `Limits(0, 2)`. But we can consider `frac(x)` as `x->oo` is `Limits(0, 1)`.

jksuom replied:
I forgot to divide by 2.

jksuom replied:
> Did i missed something?

We should not think of ``Limits(0, 1)`` as representing a *fixed* number ``a`` in the interval [0. 1].
Instead, we should keep in mind that it is used in connection with a function, as the *set* of its accumulation values (possibly coming arbitrarily close to 1).

This is quite analogous to ``1**oo``. In the docstring of ``Pow`` there is
>     | 1**oo        | nan     | Because there are various cases where         |
    | 1**-oo       |         | lim(x(t),t)=1, lim(y(t),t)=oo (or -oo),       |
    | 1**zoo       |         | but lim( x(t)**y(t), t) != 1.  See [3].       |
 
Only, in this case, we are able to give a more precise result: ``Limits(0, oo)``. (In fact, that could also be the value of ``1**oo``.)

gxyd replied:
> We should not think of Limits(0, 1) as representing a fixed number a in the interval [0. 1].

Yes, i know that.
> This is quite analogous to 1**oo. In the docstring of Pow there is
    | 1**oo        | nan     | Because there are various cases where         |
    | 1**-oo       |         | lim(x(t),t)=1, lim(y(t),t)=oo (or -oo),       |
    | 1**zoo       |         | but lim( x(t)**y(t), t) != 1.  See [3].       |
Only, in this case, we are able to give a more precise result: Limits(0, oo). (In fact, that could also be the value of 1**oo.)

Perhaps it seems i missed something from your second point. Thanks

jksuom commented on sympy/calculus/util.py:
```diff
@@ -112,3 +119,672 @@ def elm_domain(expr, intrvl):
+        Returns True if range of values attained by `self` Limits object
+        is greater than the range of values attained by `other`, where other
+        may be any value of type Limits object or extended real number value,
+        , False is returned if `other` satisfies the same property, None if
```
Typo: comma and space.

jksuom commented on sympy/calculus/util.py:
```diff
@@ -112,3 +119,672 @@ def elm_domain(expr, intrvl):
+        >>> from sympy import Limits, oo
+        >>> 1 in Limits(-1, 3)
+        True
+        >>> -oo in Limits(1, oo)
```
This should be changed. One should always think that ``-oo`` and ``oo`` go together as limits.

jksuom commented on sympy/functions/elementary/trigonometric.py:
```diff
@@ -855,11 +874,27 @@ def inverse(self, argindex=1):
+                return Limits(S.NegativeInfinity, S.Infinity)
+
+        if isinstance(arg, Limits):
+            min, max = arg.min, arg.max
```
It appears that these values are not actually used.

gxyd commented:
I have addressed the comments. Also `frac(oo)` now returns `⟨0,∞⟩`.

jksuom commented on sympy/functions/elementary/trigonometric.py:
```diff
@@ -855,11 +874,27 @@ def inverse(self, argindex=1):
+                min = min - floor(min/S.Pi)*S.Pi
+
+            if max is not S.Infinity:
+                max = max - floor(min/S.Pi)*S.Pi
```
I think that ``floor(min/S.Pi)`` should have the same value as above. Perhaps one should first set
`` d = floor(min/S.Pi)*S.Pi``
and then ``min -= d``, ``max -= d``.

gxyd commented:
For your comment on `-oo in Limits(1, oo)` returning `True`. I just made the change blindly for what you said, may be there is some crucial part of `Limits` that i am missing?
I looked upon docstring again and made some changes in the latest commit.

jksuom commented on sympy/calculus/util.py:
```diff
@@ -112,3 +119,669 @@ def elm_domain(expr, intrvl):
+    >>> log(Limits(1, E))
+    <0, 1>
+
+    Some symbol in an experession can be substituted for an Interval.
```
experession -> expression

jksuom commented on sympy/calculus/util.py:
```diff
@@ -112,3 +119,669 @@ def elm_domain(expr, intrvl):
+
+    `\langle a, b\rangle = \{x \in \mathbb{R} \mid a \le x \le b\}`
+
+    `\langle -\infty, b\rangle = \{x \in \mathbb{R} \cup \{-\infty\} \mid x \le b\}`
```
I think ``oo`` should be added to this. This form
``\langle -\infty, b\rangle = \{x \in \mathbb{R} \mid x \le b\} \cup \{-\infty, \infty}``
might look best. That is, if either ``-oo`` or ``oo`` is an argument, then the other one should be included (though not as an end point).
This is forced, since we have, for example, ``1/Limits(0, 1) == Limits(1, oo)``, and the limit at 0 is not one-sided. As ``x < 0`` tends to 0, then ``1/x -> -oo``, so ``-oo`` should be interpreted as belonging to ``Limits(1, oo)`` though it does not (and need not) appear explicitly.

gxyd replied:
Now it much more clear to me. :)

gxyd commented:
Currently
```
>>> ((tan(x))**(sin(2*x))).subs({x: Limits(0, pi/2)})
gxyd@swap:~/Public/sympy$ python3.4
Python 3.4.3 (default, Oct 14 2015, 20:28:29) 
[GCC 4.8.4] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from sympy import *
>>> x, y, z = symbols('x y z')
>>> ((tan(x))**(sin(2*x))).subs({x: Limits(0, pi/2)})
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/gxyd/Public/sympy/sympy/core/basic.py", line 895, in subs
    rv = rv._subs(old, new, **kwargs)
  File "/home/gxyd/Public/sympy/sympy/core/cache.py", line 93, in wrapper
    retval = cfunc(*args, **kwargs)
  File "/usr/lib/python3.4/functools.py", line 472, in wrapper
    result = user_function(*args, **kwds)
  File "/home/gxyd/Public/sympy/sympy/core/basic.py", line 1009, in _subs
    rv = fallback(self, old, new)
  File "/home/gxyd/Public/sympy/sympy/core/basic.py", line 986, in fallback
    rv = self.func(*args)
  File "/home/gxyd/Public/sympy/sympy/core/cache.py", line 93, in wrapper
    retval = cfunc(*args, **kwargs)
  File "/usr/lib/python3.4/functools.py", line 472, in wrapper
    result = user_function(*args, **kwds)
  File "/home/gxyd/Public/sympy/sympy/core/power.py", line 192, in __new__
    obj = b._eval_power(e)
  File "/home/gxyd/Public/sympy/sympy/core/decorators.py", line 77, in __sympifyit_wrapper
    return func(a, b)
  File "/home/gxyd/Public/sympy/sympy/calculus/util.py", line 318, in _eval_power
    return self.__pow__(other)
  File "/home/gxyd/Public/sympy/sympy/core/decorators.py", line 77, in __sympifyit_wrapper
    return func(a, b)
  File "/home/gxyd/Public/sympy/sympy/calculus/util.py", line 546, in __pow__
    num_pow = self**num
  File "/home/gxyd/Public/sympy/sympy/core/decorators.py", line 77, in __sympifyit_wrapper
    return func(a, b)
  File "/home/gxyd/Public/sympy/sympy/calculus/util.py", line 546, in __pow__
    num_pow = self**num
  File "/home/gxyd/Public/sympy/sympy/core/decorators.py", line 77, in __sympifyit_wrapper
    return func(a, b)
  File "/home/gxyd/Public/sympy/sympy/calculus/util.py", line 546, in __pow__
    num_pow = self**num
  File "/home/gxyd/Public/sympy/sympy/core/decorators.py", line 77, in __sympifyit_wrapper
    return func(a, b)
  File "/home/gxyd/Public/sympy/sympy/calculus/util.py", line 546, in __pow__
    num_pow = self**num
  File "/home/gxyd/Public/sympy/sympy/core/decorators.py", line 77, in __sympifyit_wrapper
    return func(a, b)
  File "/home/gxyd/Public/sympy/sympy/calculus/util.py", line 546, in __pow__
    num_pow = self**num
  File "/home/gxyd/Public/sympy/sympy/core/decorators.py", line 77, in __sympifyit_wrapper
    return func(a, b)
  File "/home/gxyd/Public/sympy/sympy/calculus/util.py", line 546, in __pow__
    num_pow = self**num
  File "/home/gxyd/Public/sympy/sympy/core/decorators.py", line 77, in __sympifyit_wrapper
    return func(a, b)
  File "/home/gxyd/Public/sympy/sympy/calculus/util.py", line 546, in __pow__
    num_pow = self**num
  File "/home/gxyd/Public/sympy/sympy/core/decorators.py", line 77, in __sympifyit_wrapper
    return func(a, b)
  File "/home/gxyd/Public/sympy/sympy/calculus/util.py", line 546, in __pow__
    num_pow = self**num
  File "/home/gxyd/Public/sympy/sympy/core/decorators.py", line 77, in __sympifyit_wrapper
    return func(a, b)
  File "/home/gxyd/Public/sympy/sympy/calculus/util.py", line 546, in __pow__
    num_pow = self**num
  File "/home/gxyd/Public/sympy/sympy/core/decorators.py", line 77, in __sympifyit_wrapper
    return func(a, b)
  File "/home/gxyd/Public/sympy/sympy/calculus/util.py", line 546, in __pow__
    num_pow = self**num
  File "/home/gxyd/Public/sympy/sympy/core/decorators.py", line 77, in __sympifyit_wrapper
    return func(a, b)
  File "/home/gxyd/Public/sympy/sympy/calculus/util.py", line 546, in __pow__
    num_pow = self**num
  File "/home/gxyd/Public/sympy/sympy/core/decorators.py", line 77, in __sympifyit_wrapper
    return func(a, b)
  File "/home/gxyd/Public/sympy/sympy/calculus/util.py", line 546, in __pow__
    num_pow = self**num
  File "/home/gxyd/Public/sympy/sympy/core/decorators.py", line 77, in __sympifyit_wrapper
    return func(a, b)
  File "/home/gxyd/Public/sympy/sympy/calculus/util.py", line 546, in __pow__
    num_pow = self**num
  File "/home/gxyd/Public/sympy/sympy/core/decorators.py", line 77, in __sympifyit_wrapper
    return func(a, b)
  File "/home/gxyd/Public/sympy/sympy/calculus/util.py", line 546, in __pow__
    num_pow = self**num
  File "/home/gxyd/Public/sympy/sympy/core/decorators.py", line 77, in __sympifyit_wrapper
    return func(a, b)
  File "/home/gxyd/Public/sympy/sympy/calculus/util.py", line 546, in __pow__
    num_pow = self**num
  File "/home/gxyd/Public/sympy/sympy/core/decorators.py", line 77, in __sympifyit_wrapper
    return func(a, b)
  File "/home/gxyd/Public/sympy/sympy/calculus/util.py", line 546, in __pow__
    num_pow = self**num
  File "/home/gxyd/Public/sympy/sympy/core/decorators.py", line 77, in __sympifyit_wrapper
    return func(a, b)
  File "/home/gxyd/Public/sympy/sympy/calculus/util.py", line 546, in __pow__
    num_pow = self**num
  File "/home/gxyd/Public/sympy/sympy/core/decorators.py", line 77, in __sympifyit_wrapper
    return func(a, b)
  File "/home/gxyd/Public/sympy/sympy/calculus/util.py", line 546, in __pow__
    num_pow = self**num
  File "/home/gxyd/Public/sympy/sympy/core/decorators.py", line 77, in __sympifyit_wrapper
    return func(a, b)
  File "/home/gxyd/Public/sympy/sympy/calculus/util.py", line 546, in __pow__
    num_pow = self**num
  File "/home/gxyd/Public/sympy/sympy/core/decorators.py", line 77, in __sympifyit_wrapper
    return func(a, b)
  File "/home/gxyd/Public/sympy/sympy/calculus/util.py", line 546, in __pow__
    num_pow = self**num
  File "/home/gxyd/Public/sympy/sympy/core/decorators.py", line 77, in __sympifyit_wrapper
    return func(a, b)
  File "/home/gxyd/Public/sympy/sympy/calculus/util.py", line 546, in __pow__
    num_pow = self**num
  File "/home/gxyd/Public/sympy/sympy/core/decorators.py", line 77, in __sympifyit_wrapper
    return func(a, b)
  File "/home/gxyd/Public/sympy/sympy/calculus/util.py", line 546, in __pow__
    num_pow = self**num
  File "/home/gxyd/Public/sympy/sympy/core/decorators.py", line 77, in __sympifyit_wrapper
    return func(a, b)
  File "/home/gxyd/Public/sympy/sympy/calculus/util.py", line 546, in __pow__
    num_pow = self**num
  File "/home/gxyd/Public/sympy/sympy/core/decorators.py", line 77, in __sympifyit_wrapper
    return func(a, b)
  File "/home/gxyd/Public/sympy/sympy/calculus/util.py", line 546, in __pow__
    num_pow = self**num
  File "/home/gxyd/Public/sympy/sympy/core/decorators.py", line 77, in __sympifyit_wrapper
    return func(a, b)
  File "/home/gxyd/Public/sympy/sympy/calculus/util.py", line 546, in __pow__
    num_pow = self**num
  File "/home/gxyd/Public/sympy/sympy/core/decorators.py", line 77, in __sympifyit_wrapper
    return func(a, b)
  File "/home/gxyd/Public/sympy/sympy/calculus/util.py", line 546, in __pow__
    num_pow = self**num
  File "/home/gxyd/Public/sympy/sympy/core/decorators.py", line 77, in __sympifyit_wrapper
    return func(a, b)
  File "/home/gxyd/Public/sympy/sympy/calculus/util.py", line 546, in __pow__
    num_pow = self**num
  File "/home/gxyd/Public/sympy/sympy/core/decorators.py", line 77, in __sympifyit_wrapper
    return func(a, b)
  File "/home/gxyd/Public/sympy/sympy/calculus/util.py", line 546, in __pow__
    num_pow = self**num
  File "/home/gxyd/Public/sympy/sympy/core/decorators.py", line 77, in __sympifyit_wrapper
    return func(a, b)
  File "/home/gxyd/Public/sympy/sympy/calculus/util.py", line 546, in __pow__
    num_pow = self**num
  File "/home/gxyd/Public/sympy/sympy/core/decorators.py", line 77, in __sympifyit_wrapper
    return func(a, b)
  File "/home/gxyd/Public/sympy/sympy/calculus/util.py", line 546, in __pow__
    num_pow = self**num
  File "/home/gxyd/Public/sympy/sympy/core/decorators.py", line 77, in __sympifyit_wrapper
    return func(a, b)
  File "/home/gxyd/Public/sympy/sympy/calculus/util.py", line 546, in __pow__
    num_pow = self**num
  File "/home/gxyd/Public/sympy/sympy/core/decorators.py", line 77, in __sympifyit_wrapper
    return func(a, b)
  File "/home/gxyd/Public/sympy/sympy/calculus/util.py", line 546, in __pow__
    num_pow = self**num
  File "/home/gxyd/Public/sympy/sympy/core/decorators.py", line 77, in __sympifyit_wrapper
    return func(a, b)
  File "/home/gxyd/Public/sympy/sympy/calculus/util.py", line 546, in __pow__
    num_pow = self**num
  File "/home/gxyd/Public/sympy/sympy/core/decorators.py", line 77, in __sympifyit_wrapper
    return func(a, b)
  File "/home/gxyd/Public/sympy/sympy/calculus/util.py", line 546, in __pow__
    num_pow = self**num
  File "/home/gxyd/Public/sympy/sympy/core/decorators.py", line 77, in __sympifyit_wrapper
    return func(a, b)
  File "/home/gxyd/Public/sympy/sympy/calculus/util.py", line 546, in __pow__
    num_pow = self**num
  File "/home/gxyd/Public/sympy/sympy/core/decorators.py", line 77, in __sympifyit_wrapper
    return func(a, b)
  File "/home/gxyd/Public/sympy/sympy/calculus/util.py", line 546, in __pow__
    num_pow = self**num
  File "/home/gxyd/Public/sympy/sympy/core/decorators.py", line 77, in __sympifyit_wrapper
    return func(a, b)
  File "/home/gxyd/Public/sympy/sympy/calculus/util.py", line 546, in __pow__
    num_pow = self**num
  File "/home/gxyd/Public/sympy/sympy/core/decorators.py", line 77, in __sympifyit_wrapper
    return func(a, b)
  File "/home/gxyd/Public/sympy/sympy/calculus/util.py", line 546, in __pow__
    num_pow = self**num
  File "/home/gxyd/Public/sympy/sympy/core/decorators.py", line 77, in __sympifyit_wrapper
    return func(a, b)
  File "/home/gxyd/Public/sympy/sympy/calculus/util.py", line 546, in __pow__
    num_pow = self**num
  File "/home/gxyd/Public/sympy/sympy/core/decorators.py", line 77, in __sympifyit_wrapper
    return func(a, b)
  File "/home/gxyd/Public/sympy/sympy/calculus/util.py", line 546, in __pow__
    num_pow = self**num
  File "/home/gxyd/Public/sympy/sympy/core/decorators.py", line 77, in __sympifyit_wrapper
    return func(a, b)
  File "/home/gxyd/Public/sympy/sympy/calculus/util.py", line 546, in __pow__
    num_pow = self**num
  File "/home/gxyd/Public/sympy/sympy/core/decorators.py", line 77, in __sympifyit_wrapper
    return func(a, b)
  File "/home/gxyd/Public/sympy/sympy/calculus/util.py", line 546, in __pow__
    num_pow = self**num
  File "/home/gxyd/Public/sympy/sympy/core/decorators.py", line 77, in __sympifyit_wrapper
    return func(a, b)
  File "/home/gxyd/Public/sympy/sympy/calculus/util.py", line 546, in __pow__
    num_pow = self**num
  File "/home/gxyd/Public/sympy/sympy/core/decorators.py", line 77, in __sympifyit_wrapper
    return func(a, b)
  File "/home/gxyd/Public/sympy/sympy/calculus/util.py", line 546, in __pow__
    num_pow = self**num
  File "/home/gxyd/Public/sympy/sympy/core/decorators.py", line 77, in __sympifyit_wrapper
    return func(a, b)
  File "/home/gxyd/Public/sympy/sympy/calculus/util.py", line 546, in __pow__
    num_pow = self**num
  File "/home/gxyd/Public/sympy/sympy/core/decorators.py", line 77, in __sympifyit_wrapper
    return func(a, b)
  File "/home/gxyd/Public/sympy/sympy/calculus/util.py", line 540, in __pow__
    if num == S(1):
  File "/home/gxyd/Public/sympy/sympy/core/sympify.py", line 253, in sympify
    return converter[cls](a)
  File "/home/gxyd/Public/sympy/sympy/core/numbers.py", line 1667, in __new__
    if isinstance(i, string_types):
RuntimeError: maximum recursion depth exceeded while calling a Python object
```
The latest commit fixes this.

gxyd commented:
I have made the changes in name of `Limits` -> `AccumulationBounds`(alias `AccumBounds`). I think keeping both the names is a better option.

gxyd commented:
@jksuom does the current name changes look right? Perhaps only `PyPy` failures(seem un-related).

jksuom commented:
The name changes look OK (though I haven't reviewed everything yet). I have restarted the PyPy job.

jksuom commented on sympy/calculus/util.py:
```diff
@@ -112,3 +119,698 @@ def elm_domain(expr, intrvl):
+
+    otherwise
+
+    `X^n = \{ x^n \mid x \neq 0, x \in X\} \cup \{-\infty, \infty\}`
```
Is this for the case ``n < 0``?

gxyd replied:
Acknowledged

jksuom commented on sympy/calculus/util.py:
```diff
@@ -112,3 +119,698 @@ def elm_domain(expr, intrvl):
+    =====
+
+    Do not use ``AccumulationBounds`` for floating point interval arithmetic
+    calculations use ``mpmath.iv`` instead.
```
calculations, use ...

jksuom commented on sympy/calculus/util.py:
```diff
@@ -112,3 +119,698 @@ def elm_domain(expr, intrvl):
+
+        return And(self.min <= other and self.max >= other)
+
+    def intersect(self, other):
```
Perhaps this could be ``intersection`` (a noun).

gxyd commented:
Though i have addressed the comments. Also i will see the coverage report and add more tests if i find its need.

gxyd commented:
@jksuom here is percentage 
![report_calculus](https://cloud.githubusercontent.com/assets/7712241/11849123/330b74c0-a44c-11e5-985d-13050bb515dd.png)
have a look on [coverage report for util.py(only)](https://medium.com/@gauravdhingra/accumulationbounds-coverage-in-util-py-only-78df9db28067#.jzh1745aj) . Coverage of other files like `trigonometry` and other remaining is good.

certik commented:
@gxyd thanks for the contribution. What is the status of this? Do you plan to add more stuff or do you want this merged?

gxyd commented:
I think you can leave merging task upto @jksuom , if he has any comments i will address those. Otherwise it is ready to merge. Thanks @certik

jksuom commented:
@gxyd I think this is ready now. You may want to do some squashing before merge?

gxyd commented:
I have done the squashing in appropriate commits. Also i forgot mention the PR #2682 (for future rerefnece work of `ComplexAccumulationBounds` may be) by @hargup from which i took the idea of operations, though i mentioned it in commit.

gxyd commented:
Also for [this comment](https://github.com/sympy/sympy/pull/10051#issuecomment-160053689) , i have added tests for what they "will" return after PR is merged. So i think the corresponding two issues can also be closed. I will mention the two issues in my commit to close them.

jksuom commented on sympy/printing/codeprinter.py:
```diff
@@ -492,6 +492,7 @@ def _print_not_supported(self, expr):
     _print_Infinity = _print_not_supported
     _print_Integral = _print_not_supported
     _print_Interval = _print_not_supported
+    _print_Limits = _print_not_supported
```
Maybe this should be changed?

gxyd replied:
Ya, i have changed it. No other similar error is there i verified.

jksuom commented:
Thanks! I'll merge when tests pass.

hargup commented:
@gxyd Thanks for giving me credit. From reading the docstring and tests the PR looks pretty good. I suggest that you move `AccumulationBounds` to a separate file. You should also update the documentation of functions like `sin` to reflect the fact that they can now accept and return an `AccumulationBound`.

jksuom commented:
Yes, it seems that the class has grown bigger than just an "util". And there remain other tasks such as additions to documentation. But I feel that that could wait until we get more experienced with the implications.

gxyd commented:
Thanks @jksuom @hargup . I will try to see the implementation of ComplexAccumulationBounds in a few months will also make the changes suggested by @hargup

asmeurer commented:
How would complex accumulation bounds work?

asmeurer commented:
Also note https://en.m.wikipedia.org/wiki/Picard_theorem. "Well behaved" complex functions take on (almost) every value across the complex plane.

jksuom commented:
I'm also afraid that complex accumulation bounds may not be very useful. In fact, the strong form of Picard's theorem says that the set of accumulation values of a meromorphic function in a neighbourhood of an essential singularity is the whole extended complex plane. (This will also follow from Casorati-Weierstrass theorem that is much easier to prove.) Other singularities are just poles where the function has limit ``zoo`` (or entirely removable, cf. #8320).
